### PR TITLE
Add gcOp feature to document schema

### PR DESCRIPTION
**Pull request by markfields**
## Description

Fixes [AB#7928](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/7928)

Rather than using config overrides (aka feature gates) to disable GC at first for 2.0, use a ContainerRuntime option which AzureClient/OdspClient set based on compatibility requirements.  i.e. AzureClient _does_ set the new option `blockGCOpForSchemaCompatibility` because by default 1.0/2.0 clients should play nice, but OdspClient does not, since all clients will be v2+.

## Breaking Changes

No. The default behavior is unchanged, which is to say: If GC Sweep is otherwise enabled, it will work as designed including submitting the GC op.

## Reviewer Guidance

* Many of the changed files are 1-liners in test files due to updated params objects
* Still need to finish cleaning up / adding tests.

### CODE REVIEW

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 1]
```diff
@@ -976,6 +976,14 @@ export class ContainerRuntime
 			compressionOptions.minimumBatchSizeInBytes !== Infinity &&
 			compressionOptions.compressionAlgorithm === "lz4";
 
+		// If both these settings are set this way, we will certainly not submit the GC op.
+		// In this case, we declare in Document Schema below that the gcOp feature is not supported.
+		// Once any client stops blocking GC op for a document, this schema bit will flip
+		// and clients that don't understand or unblock the GC op will face container closure on boot.
+		const gcOpBlocked =
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/containerRuntime.ts:983]: I would not do this here. Document schema should only tell whether GC ops is supported or not. This is passed to garbage collector which decides whether to send GC ops or not. That is a simpler model and is consistent with how other features are used.

Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts:983]: This should be done based on the `shouldRunSweep` flag. If that is enabled, we could send GC op and clients that do not understand this op should close.

Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts:983]: We can enable sweep + tombstone based on `shouldRunSweep`

Comment (markfields) [reply packages/runtime/container-runtime/src/containerRuntime.ts]: We paired on a better solution.

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 2]
```diff
@@ -417,6 +417,16 @@ export interface ICompressionRuntimeOptions {
...
 	readonly gcOptions?: IGCRuntimeOptions;
+
+	/**
+	 * When upgrading to 2.0, application authors need to defer the schema upgrade implicit in sending the GC op.
+	 * So provide a way to block the GC op (and all related features - GC Sweep/Tombstone) until ready for schema upgrade.
+	 *
+	 * If not true, the document schema will express support for GC op (updating if needed),
+	 * and any clients that don't support it will close on boot (e.g. 1.x clients or those still setting this to true)
+	 */
+	readonly blockGCOpForSchemaCompatibility?: boolean;
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:428]: So, default value means GC is on, right?
It's a bit hard to parse both comment and name of the property, maybe there is a better way to name it?

File: packages/runtime/container-runtime/src/summary/documentSchema.ts [hunk 3]
```diff
@@ -95,6 +95,11 @@ export interface IDocumentSchemaFeatures {
 	compressionLz4: boolean;
 	idCompressorMode: IdCompressorMode;
 	opGroupingEnabled: boolean;
+	/**
+	 * Is the GC op able to be processed?
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/summary/documentSchema.ts:99]: Maybe - is GC ops allowed / part of document schema?

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 2 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/containerRuntime.ts:428]: Probably so :) open to suggestions. Was trying to be extra verbose / explicit while reasoning over it.

File: packages/runtime/container-runtime/src/summary/documentSchema.ts [hunk 4]
```diff
@@ -219,6 +224,7 @@ const documentSchemaSupportedConfigs = {
 	explicitSchemaControl: new TrueOrUndefinedMax(), // once new behavior shows up, it's sticky
 	idCompressorMode: new IdCompressorProperty(["delayed", "on"]),
 	opGroupingEnabled: new TrueOrUndefined(),
+	gcOpSupported: new TrueOrUndefined(),
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/summary/documentSchema.ts:227]: Why is it `gcOpSupported`, when all the other places use `gcOp`?
If that's the bug, what is missing in terms of coverage to diagnose this bug? I'd think that existing UTs should fail

Comment (markfields) [reply packages/runtime/container-runtime/src/summary/documentSchema.ts:227]: Oops yep something should fail

Comment (markfields) [reply packages/runtime/container-runtime/src/summary/documentSchema.ts:227]: Some unit tests failed. I have a concept of how we can get compile-time protection (or remove redundant declaration of type and value), I'll try it in another PR if I have time

**Comment by markfields** [comment]
We are not going to account for the GC op in Document Schema after all, for two reasons:
- It is very challenging to retrofit a feature like this without disrupting the production scenarios that depend on GC op.  I.e. We could have successfully used Document Schema from the start, but not now, many releases later.
- 1.x/2.0 compat is already protected by other Document Schema features that are set up.
